Make admins adhere to branch protection rules

enko opened the pull request:
This introduces a new flag `BlockAdminMergeOverride` on the branch protection rules that prevents admins/repo owners from bypassing branch protection rules and merging without approvals or failing status checks.

Fixes #17131

lunny commented:
Wouldn't the admin can change the protection rules?

wxiaoguang commented on templates/repo/issue/view_content/pull.tmpl:
```diff
@@ -164,7 +164,7 @@
 				{{$notAllOverridableChecksOk := or .IsBlockedByApprovals .IsBlockedByRejection .IsBlockedByOfficialReviewRequests .IsBlockedByOutdatedBranch .IsBlockedByChangedProtectedFiles (and .EnableStatusCheck (not .RequiredStatusCheckState.IsSuccess))}}
 
 				{{/* admin can merge without checks, writer can merge when checks succeed */}}
```
```suggestion
				{{/* admin can merge without checks, writer can merge when checks succeed, BlockAdminMergeOverride only hides the button on web UI but doesn't stop the admin use API or web request to force merging */}}
```

I think this change means "UI-only hidden", but not completely prevent admin users from merging.

enko replied:
Thanks for the pointer. I'll look into changing the API endpoints for merging pull requests.

enko commented:
> Wouldn't the admin can change the protection rules?

Yes, but that would leave a paper trail. 

It is the same on GitHub.

lunny commented:
> > Wouldn't the admin can change the protection rules?
> 
> Yes, but that would leave a paper trail.
> 
> It is the same on GitHub.

I know Github's behaviour but I suspect it's not a better solution than having a configuration item in app.ini

enko commented:
> > I know Github's behaviour but I suspect it's not a better solution than having a configuration item in app.ini

I see why you would want to have it in your `app.ini` to have this setting behave very strict. My point would be that it is primarily about compliance and accidental merge. If an owner/admin changes the setting, he has to actively do it and if Gitea would have an audit log, it would leave a paper trail.

> But why a "global" config item? For large instance, different organizations have different requirements.

Yes, different projects have different needs and sometimes even different branches have different needs.

> […] Gitea does need a well-designed config system, global -> org-level -> repo-level.

That is out of scope.

enko commented:
@wxiaoguang I added the service changes and also added a test case.

wxiaoguang commented:
Thank you for the update. There are still some things left:

1. There are some lint errors.
1. Some structs in "modules/structs/repo_branch.go" are changed, but I do not see how they are used in code. If I read correctly, only the one in "ProtectBranchForm" is really used in code.
1. Database change needs a database migration.

enko commented:
@wxiaoguang Thanks for your input!

> Thank you for the update. There are still some things left:
> 
>     1. There are some lint errors.

Fixed that, sorry for that.

>     2. Some structs in "modules/structs/repo_branch.go" are changed, but I do not see how they are used in code. If I read correctly, only the one in "ProtectBranchForm" is really used in code.

I checked, and I noticed I missed the API did not use the new field. I added that.

>     3. Database change needs a database migration.

Added a migration, I hope I did that right?

wxiaoguang commented on services/pull/check.go:
```diff
@@ -118,13 +118,26 @@ func CheckPullMergeable(stdCtx context.Context, doer *user_model.User, perm *acc
-				if isRepoAdmin, errCheckAdmin := access_model.IsUserRepoAdmin(ctx, pr.BaseRepo, doer); errCheckAdmin != nil {
+				pb, pbErr := git_model.GetFirstMatchProtectedBranchRule(ctx, pr.BaseRepoID, pr.BaseBranch)
+				if pbErr != nil {
+					return fmt.Errorf("LoadProtectedBranch: %v", err)
```
It seems not right. It should use `pbErr` but not `err`, and `Errorf` should use `%w` to wrap the error.

I will make some changes.

wxiaoguang commented:
Some changes in 71b5edd500b282d312ca068679d7f41b934d86a4:

1. Make the database migration work `models/migrations/migrations.go`)
1. Add some spaces in `options/locale/locale_en-US.ini`
1. Use `Errorf(%w)` to wrap the correct errors
1. Simplify some `if` blocks
1. Rename the argument `adminSkipProtectionCheck` to `adminForceMerge` to match the callers `form.ForceMerge`.

wxiaoguang commented on templates/repo/issue/view_content/pull.tmpl:
```diff
@@ -164,7 +164,7 @@
 
 				{{/* admin can merge without checks, writer can merge when checks succeed */}}
-				{{$canMergeNow := and (or $.IsRepoAdmin (not $notAllOverridableChecksOk)) (or (not .AllowMerge) (not .RequireSigned) .WillSign)}}
+				{{$canMergeNow := and (or (and (not $.BlockAdminMergeOverride) $.IsRepoAdmin) (not $notAllOverridableChecksOk)) (or (not .AllowMerge) (not .RequireSigned) .WillSign)}}
```
It seems that the red "merge" button is still on the UI?

![image](https://github.com/user-attachments/assets/5ed3d083-fd29-41f7-ba4c-6d968e6db5fa)

lalo-mx commented on models/migrations/migrations.go:
```diff
@@ -603,6 +603,8 @@ var migrations = []Migration{
 	// v305 -> v306
 	NewMigration("Add Repository Licenses", v1_23.AddRepositoryLicenses),
+	// v306 -> v307
+	NewMigration("Add Repository Licenses", v1_23.AddBlockAdminMergeOverrideBranchProtection),
```
Migration title is repeated from last migration

wxiaoguang replied:
```suggestion
	NewMigration("Add BlockAdminMergeOverride to ProtectedBranch", v1_23.AddBlockAdminMergeOverrideBranchProtection),
```

Oooooooooops, I must be not awaken .....

lunny commented:
<img width="999" alt="图片" src="https://github.com/user-attachments/assets/806faa27-3717-460c-9b71-faa03e904b6d">
Scheduled merge should also be checked from template.

wxiaoguang commented:
> Scheduled merge should also be checked from template.

Why?


![image](https://github.com/user-attachments/assets/acc71e0d-0a85-4446-ac90-af6fc9ccd0fb)

lunny commented:
I mean auto-merge means should be hidden for the administrator from the UI.

wxiaoguang commented:
> I mean auto-merge means should be hidden for the administrator from the UI.

Why they should hide? Auto merge only merge when checks pass.

lunny commented:
> > I mean auto-merge means should be hidden for the administrator from the UI.
> 
> Why they should hide? Auto merge only merge when checks pass.

Yes, but the permissions that can merge the pull request are checked when creating the schedule.

wxiaoguang commented:
> > > I mean auto-merge means should be hidden for the administrator from the UI.
> > 
> > 
> > Why they should hide? Auto merge only merge when checks pass.
> 
> Yes, but the permissions that can merge the pull request are checked when creating the schedule.

I do not understand what you mean. Show a real case, what's wrong would happen.

lunny commented:
> @wxiaoguang @lunny Thank you for your support in bringing this feature to fruition.
> 
> Is there anything for me to do?

No action is needed unless additional maintainers submit new review requests. Otherwise, this pull request will be merged in a few days.
